### PR TITLE
use faster timers

### PR DESCRIPTION
**Pull request by ronag**
Another major bottleneck I found is the constant refreshing of timers, in particular the body timeout timer which is refresh every time we get a packet.

Our current use of timers is that we have a very few timers which we update frequently where high accuracy is not required. This PR tries to implement that.

### CODE REVIEW

**Comment by ronag** [comment]
@mcollina This breaks all the tests using fake timers... thoughts on what to do about that?

**Comment by mcollina** [comment]
What benefits is this giving us?

I would prefer this timer to be attached to the agent, so we can actually stop them. If this is significantly faster, why aren't we briging that optimization to Node.js itself?

**Comment by ronag** [comment]
> What benefits is this giving us?

In high throughput scenarios with lots of small packets we are constantly calling `refresh` on node timers which is much more expensive than what this PR does.

> I would prefer this timer to be attached to the agent, so we can actually stop them. If this is significantly faster, why aren't we briging that optimization to Node.js itself?

I don't follow... Node has a more generic timer implementation which works for other less nisch use cases.

**Comment by ronag** [comment]
@mcollina PTAL

**Comment by ronag** [comment]
Note that the timers in this PR are *much* less accurate than note timers, which I believe is fine for our use case.

**Comment by mcollina** [comment]
Have you got a benchmark before/after?

**Comment by ronag** [comment]
Before:

[bench:run] │ undici - stream     │       1 │ 7383.29 req/sec │  ± 0.00 % │              + 395.85 % │

After:

[bench:run] │ undici - dispatch   │       1 │ 8045.52 req/sec │  ± 0.00 % │              + 380.21 % │

File: lib/client.js [hunk 1]
```diff
@@ -84,6 +84,35 @@ try {
   channels.connected = { hasSubscribers: false }
 }
 
+let fastNow = Date.now()
+const fastTimers = []
+const fastNowInterval = setInterval(() => {
```
Comment (anonrig) [inline lib/client.js:89]: When should we call clearInterval on this?

**Comment by mcollina** [comment]
Note this is the same bottleneck of the slowloris mitigation we have in core. Works kind of the same too.

**Comment by mcollina** [comment]
cc @ShogunPanda

**Comment by ronag** [comment]
> Note this is the same bottleneck of the slowloris mitigation we have in core. Works kind of the same too.

slowloris?

**Comment by KhafraDev** [comment]
> in particular the body timeout timer which is refresh every time we get a packet.

Rather than refreshing the timer, or making the timers less reliable (although not particularly important), it might be possible to let the timer run down, and re-calculate the delay based on the last packet received, or throw an error when the timer ends. Not sure if that sounds stupid or not lol

```js
lastPacketReceived = Date.now() // set on every packet

let timer = setTimeout(onBodyTimeout, bodyTimeout)

function onBodyTimeout () {
  if (Date.now() - lastPacketReceived >= bodyTimeout)
    throw ...

  timer = setTimeout(onBodyTimeout, bodyTimeout - (Date.now() - lastPacketReceived))
}

File: lib/client.js [hunk 2]
```diff
@@ -84,6 +84,35 @@ try {
...
+      len = fastTimers.length
+    } else {
+      if (timer.timeoutExpires && fastNow >= timer.timeoutExpires) {
+        timer.timeoutExpires = 0
+        onParserTimeout(timer)
+      }
+      idx += 1
+    }
+  }
+}, 1e3)
```
Comment (mcollina) [inline lib/client.js:111]: This should be active only if there is an http going on. Given how this is implemented, every node process would get this timer.

**Review by mcollina — CHANGES_REQUESTED** [review]

**Comment by ShogunPanda** [comment]
I wonder. Wouldn't be beneficial for undici to use the same connectionsCheckingInterval I introduced for node?
So we only have one timer on the server side and one on the client side.

**Comment by ronag** [comment]
> I wonder. Wouldn't be beneficial for undici to use the same connectionsCheckingInterval I introduced for node? So we only have one timer on the server side and one on the client side.

Compared to the implementation of this PR? How is that better? Also, this would require node to expose it somehow publicly.

**Comment by ronag** [comment]
I haven't looked into connectionsCheckingInterval so much. But would an alternative be the other way around? Go from a c++ version to an inaccurate but fast js version?

**Comment by ShogunPanda** [comment]
Well, I guess you could try and measure the impact.

For you to consider: connectionsCheckingInterval is a mechanism that every x seconds check the entire pool of opened inbound connections comparing (in the same "inaccurate" way you did here) if the timeout as gone off.
He's written in C++ (leveraging C sorted sets and thus able to easily bail out early in the loop) but I think it can be easily adapted in JS. 

The only pain is to maintain the list of opened/active connections (especially when we think about KeepAlive)

**Comment by ronag** [comment]
I see. I wouldn't mind using that per se. I don't know how that would practically work in undici which is developed outside of core.

**Comment by ronag** [comment]
CI looks ok 👍

File: lib/timers.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    }
+  }
+
+  if (fastTimers.length > 0) {
+    refreshTimeout()
+  }
+}
+
+function refreshTimeout () {
+  if (fastNowTimeout && fastNowTimeout.refresh) {
```
Comment (ronag) [inline lib/timers.js:40]:
```suggestion
  if (fastNowTimeout && fastNowTimeout.refresh) {
    fastNow = Date.now()
```

**Review by anonrig — APPROVED** [review]

File: lib/timers.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    fastNow = Date.now()
```
Comment (ronag) [inline lib/timers.js:41]:
```suggestion
```

File: lib/timers.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+    this.refresh()
+  }
+
+  refresh () {
+    if (!this.active) {
+      this.active = true
+      fastTimers.push(this)
+      if (!fastNowTimeout || fastTimers.length === 1) {
+        refreshTimeout()
```
Comment (ronag) [inline lib/timers.js:68]:
```suggestion
        refreshTimeout()
        fastNow = Date.now()
```

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
CI is still very red :(

**Comment by ronag** [comment]
I will fix that

**Comment by ronag** [comment]
Seems ok now. Same as usual with windows on node 19.

**Comment by KhafraDev** [comment]
v12 CI failed on this test

```
     # Subtest: keep-alive threshold
        1..2
        ok 1 - should not error
        not ok 2 - (unnamed test)
          ---
          at:
            line: 217
            column: 13
            file: test/client-keep-alive.js
            type: Timeout
            method: _onTimeout
          stack: |
            Timeout._onTimeout (test/client-keep-alive.js:217:13)
          source: |2
                    const timeout = setTimeout(() => {
                      t.fail()
            ------------^
                    }, 2e3)
                    client.on('disconnect', () => {
          ...
        
        # failed 1 of 2 tests
    not ok 6 - keep-alive threshold # time=2017.2ms
```

**Pull request merged**